cmd/av/stack_sync.go

```diff
@@ -152,64 +256,61 @@ base branch.
+			Name: state.OriginalBranch,
+		}); err != nil {
+			return err
+		}
+
 		return nil
 	},
 }
 
-// doStackSync performs the stack sync operation.
-// It returns a boolean indicating whether or not the sync was interrupted
-// (if true, the stack sync can be resumed with the --continue flag).
-// TODO[mvp]:
-//    This whole function is kind of a hot mess, and a lot of it is rooted in
-//    how awkward the API is for dealing with "branch trees" (terminology is
-//    hard when you're creating a meta-tree-structure on top of git branches).
-//    Let's refactor that a bit and then re-work this.
-func doStackSync(repo *git.Repo, config stackSyncConfig) (bool, error) {
-	branches, err := meta.ReadAllBranches(repo)
-	if err != nil {
-		return false, err
-	}
+type syncResult struct {
+	Conflict      bool
+	CurrentBranch string
+}
 
-	currentBranch, err := repo.CurrentBranchName()
+// Find all the ancestor branches of the given branch name and append them to
+// the given slice (in topological order: a comes before b if a is an ancestor
+// of b).
+func previousBranches(branches map[string]meta.Branch, name string) ([]string, error) {
+	current, ok := branches[name]
+	if !ok {
+		return nil, errors.Errorf("branch metadata not found for %q", name)
+	}
+	if current.Parent == "" {
+		return nil, nil
+	}
+	previous, err := previousBranches(branches, current.Parent)
 	if err != nil {
-		return false, err
+		return nil, err
 	}
+	return append(previous, current.Parent), nil
+}
 
-	for {
-		currentMeta, ok := branches[currentBranch]
-		if !ok {
-			return false, errors.Errorf("stack metadata not found for branch %q", currentBranch)
-		}
-		res, err := stacks.SyncBranch(repo, &stacks.SyncBranchOpts{
-			Parent:   currentMeta.Parent,
-			Continue: config.Continue,
-		})
+func subsequentBranches(branches map[string]meta.Branch, name string) ([]string, error) {
+	logrus.Debugf("finding subsequent branches for %q", name)
+	var res []string
+	branchMeta, ok := branches[name]
+	if !ok {
+		return nil, fmt.Errorf("branch metadata not found for %q", name)
+	}
+	if len(branchMeta.Children) == 0 {
+		return res, nil
+	}
+	for _, child := range branchMeta.Children {
+		res = append(res, child)
+		desc, err := subsequentBranches(branches, child)
 		if err != nil {
-			return false, errors.WrapIff(err, "failed to sync branch %q", currentBranch)
-		}
-		switch res.Status {
-		case stacks.SyncAlreadyUpToDate:
-			fmt.Printf("Branch %q is already up-to-date with %q\n", currentBranch, currentMeta.Parent)
-		case stacks.SyncUpdated:
-			fmt.Printf("Branch %q synchronized with %q\n", currentBranch, currentMeta.Parent)
-		case stacks.SyncConflict:
-			fmt.Printf("Branch %q has merge conflict with %q, aborting...\n", currentBranch, currentMeta.Parent)
-			if res.Hint != "" {
-				_, _ = fmt.Println(text.Indent(res.Hint, "    "))
-			}
-			return true, nil
-		default:
-			logrus.Panicf("invariant error: unknown sync result: %v", res)
+			return nil, err
 		}
```

Comment:
Good job, Git, for recognizing that this curly brace actually stayed the same. lmao